lib/Transforms/IPO/StructFieldAccessInfo.cpp

```diff
@@ -0,0 +1,343 @@
+        addFieldAccessNum(Inst, FieldLoc);
+      else if (Inst->getOpcode() == Instruction::Store) {
+        if (U == Inst->getOperand(1))
+          addFieldAccessNum(Inst, FieldLoc);
+      } else {
+        if (Inst->getOpcode() == Instruction::Call) {
+          auto *Call = dyn_cast<CallInst>(Inst);
+          auto *Func = Call->getCalledFunction();
+          if (Func && Func->arg_size() == Call->getNumArgOperands()) {
+            for (unsigned i = 0; i < Call->getNumArgOperands(); i++) {
+              if (Call->getArgOperand(i) == U)
+                addFieldAccessNum(Inst, Func, i, FieldLoc);
+            }
+          } else {
+            // Bail out if Gep is used in an indirect function or a function
+            // that has undetermined args
+            addStats(StructFieldAccessManager::DebugStats::
+                         DS_GepPassedIntoIndirectFunc);
+          }
+        } else if (Inst->getOpcode() == Instruction::Invoke) {
+          auto *Call = dyn_cast<InvokeInst>(Inst);
+          auto *Func = Call->getCalledFunction();
+          if (Func) {
+            for (unsigned i = 0; i < Call->getNumArgOperands(); i++) {
+              if (Call->getArgOperand(i) == U)
+                addFieldAccessNum(Inst, Func, i, FieldLoc);
+            }
+          } else {
+            addStats(StructFieldAccessManager::DebugStats::
+                         DS_GepPassedIntoIndirectFunc);
+          }
+        } else if (Inst->getOpcode() == Instruction::BitCast) {
+          addStats(
+              StructFieldAccessManager::DebugStats::DS_GepPassedIntoBitcast);
+        } else {
+          // TODO: Collect stats of this kind of access and add analysis later
+          addStats(StructFieldAccessManager::DebugStats::DS_GepUnknownUse,
+                   Inst->getOpcode());
+        }
+      }
+    } else if (isa<Operator>(User)) {
+      auto *Inst = dyn_cast<Operator>(User);
+      if (Inst->getOpcode() == Instruction::BitCast) {
+        addStats(StructFieldAccessManager::DebugStats::DS_GepPassedIntoBitcast);
+      } else {
+        // TODO: Collect stats of this kind of access and add analysis later
+        addStats(StructFieldAccessManager::DebugStats::DS_GepUnknownUse,
+                 Inst->getOpcode());
+      }
+    } else {
+      addStats(StructFieldAccessManager::DebugStats::
+                   DS_UserNotInstructionNorOperator);
+    }
+  }
+}
+
+void StructFieldAccessInfo::analyzeUsersOfStructValue(const Value *V) {
+  assert(V->getType()->isPointerTy() &&
+         V->getType()->getPointerElementType()->isStructTy());
+  for (auto *U : V->users()) {
+    DEBUG_WITH_TYPE(DEBUG_TYPE_IR,
+                    dbgs() << "Analyzing user of " << *V << ": " << *U << "\n");
+    if (isa<Instruction>(U)) {
+      auto *Inst = dyn_cast<Instruction>(U);
+      if (!isa<GetElementPtrInst>(Inst)) {
+        // Only support access struct through GEP for now
+        if (Inst->getOpcode() == Instruction::Call) {
```

Comment:
You can simplify the code here by creating a CallSite object so that you don't need to separately handle call and invoke.